### PR TITLE
improg : tests and memory leak patch

### PR DESCRIPTION
**Pull request by pduveau**
Complete tests and memory leak patch

### CODE REVIEW

**Comment by rgerhards** [comment]
freebsd looks like a real problem

File: tests/improg_prog_confirm.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+wait_shutdown
+content_check "program data"
+if [ ! -e $RSYSLOG_DYNNAME.stderr ]; then
+	echo $RSYSLOG_DYNNAME'.stderr missing'
+	error_exit 1
+fi
+export EXPECTED='START Received
+ACK Received
+STOP Received'
+echo "$EXPECTED" | cmp - $RSYSLOG_DYNNAME.stderr
```
Comment (rgerhards) [inline tests/improg_prog_confirm.sh]: you should replace this and the if by `cmp_exact` - I can also change it as part of the merge.

File: tests/improg_prog_confirm_killonclose.sh [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+wait_shutdown
+content_check "program data"
+if [ ! -e $RSYSLOG_DYNNAME.stderr ]; then
+	echo $RSYSLOG_DYNNAME'.stderr missing'
+	error_exit 1
+fi
+export EXPECTED='START Received
+ACK Received
+SIGTERM Received'
+echo "$EXPECTED" | cmp - $RSYSLOG_DYNNAME.stderr
```
Comment (rgerhards) [inline tests/improg_prog_confirm_killonclose.sh]: same

File: tests/improg_prog_killonclose.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+startup
+shutdown_when_empty
+wait_shutdown
+content_check "program data"
+if [ ! -e $RSYSLOG_DYNNAME.stderr ]; then
+	echo $RSYSLOG_DYNNAME'.stderr missing'
+	error_exit 1
+fi
+export EXPECTED='SIGTERM Received'
+echo "$EXPECTED" | cmp - $RSYSLOG_DYNNAME.stderr
```
Comment (rgerhards) [inline tests/improg_prog_killonclose.sh]: same

File: tests/improg_prog_simple.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+}
+'
+startup
+shutdown_when_empty
+wait_shutdown
+content_check "program data"
+if [ -e $RSYSLOG_DYNNAME.stderr ]; then 
+	echo 'improg-simul stderr:'
+	cat $RSYSLOG_DYNNAME.stderr 
+	exit 1
```
Comment (rgerhards) [inline tests/improg_prog_simple.sh]: this needs to be `error_exit 1`

File: tests/improg_prog_confirm.sh (hunk 1 above)
Comment (pduveau) [reply tests/improg_prog_confirm.sh]: Ok, I did not have a look to the function. I was convincted that cmp_exact was only for the standard out log file. I change that now. Thx

Comment (pduveau) [reply tests/improg_prog_confirm.sh]: Done in last PR

File: tests/improg_prog_confirm_killonclose.sh (hunk 2 above)
Comment (pduveau) [reply tests/improg_prog_confirm_killonclose.sh]: same

File: tests/improg_prog_simple.sh (hunk 4 above)
Comment (pduveau) [reply tests/improg_prog_simple.sh]: done in last PR

**Comment by pduveau** [comment]
@rgerhards, there are no log in docker centos7 to understand why this check failed. Can you provides more data to understand ?

**Comment by rgerhards** [comment]
> there are no log in docker centos7 to understand why this check failed. Can you provides more data to understand ?

look at the next step - it outputs the logs (because make distcheck does not do that...).

**Comment by pduveau** [comment]
@rgerhards, I do not understand what's happening with those two CI. ubuntu18-distcheck is saying that it does not find ./improg-simul.sh for the 5 tests (it's referenced as a new file in PR and as executable). Centos7 has the same problem 4 times `improg: failed to execute program './improg-simul.sh': No such file or directory`. But it found it once : `rsyslog internal message (4,-1): improg: program './improg-simul.sh' (pid 9380) did not terminate within timeout (2000 ms); killing it [v8.1904.0.master]` but in reality there are no trace of execution as you can find in ubuntu18-codecov.

**Comment by rgerhards** [comment]
@pduveau I'll have a look ASAP - give me a bit. Are all files in EXTRA_DIST?

**Comment by pduveau** [comment]
I understood what was my mistake : one file missing in EXTRA_DIST. You had the right feeling.

**Comment by rgerhards** [comment]
@pduveau thx for your persistence!

**Comment by pduveau** [comment]
@rgerhards, For sure, I'll stay closed for maintenance of my contribs and perhaps more. I'll destroy my fork of Rsyslog and rsyslog-doc as soon as you'll merge docs. There are no reasons to survive right now.

**Pull request merged**